File: 240206/두 방향 탈출 가능 여부 판별하기/determine-escapableness-with-2-ways.py
grid = []
visited = []
dxs = [1,0]
dys = [0,1]
ans = 0
###################
def dfs(x,y):
    global visited,ans

    for dx, dy in zip(dxs,dys):
        nx = x+dx
        ny = y+dy
        if can_go(nx,ny):
            visited[nx][ny] = 1
            dfs(nx,ny)
    return

def can_go(x,y):
    if not in_range(x,y):
        return False
    if grid[x][y] == 0:
        return False
    if visited[x][y]:
        return False
    return True

def in_range(x,y):
    return 0<=x<N and 0<=y<M
###################
N, M = map(int,input().split())
for _ in range(N):
    grid.append(list(map(int,input().split())))
    visited.append([False for i in range(M)])

visited[0][0] = True
dfs(0,0)

if visited[N-1][M-1] : 
    print(1)
else: 
    print(0)

File: 241204/두 방향 탈출 가능 여부 판별하기/determine-escapableness-with-2-ways.cpp
#include <iostream>
#define MAX 100
using namespace std;

int n, m;
int grid[MAX][MAX];
bool visited[MAX][MAX] = {false,};

bool inRange(int x,int y){
    return (0<=x && x<n && 0<=y && y<m);
}

bool canGo(int x,int y){
    if (!inRange(x,y))
        return false;
    if (visited[x][y])
        return false;
    if (grid[x][y]==0)
        return false;
    return true;
}

void dfs(int x, int y){
    int dx[2] = {1,0};
    int dy[2] = {0,1};
    for (int d=0;d<2;d++){
        int nx = x + dx[d];
        int ny = y + dy[d];
        if (canGo(nx,ny)){
            visited[nx][ny] = true;
            dfs(nx,ny);
        }
    }
    return;
}

int main() {

    cin >> n >> m;
    for(int i=0;i<n;i++){
        for (int j=0;j<m;j++){
            cin >> grid[i][j];
        }
    }

    visited[0][0] = true;
    dfs(0,0);
    if (visited[n-1][m-1])
        cout << 1;
    else
        cout << 0;

    return 0;
}